### ISSUE TITLE
[FIX-021] Validate integer counts before training and layer construction

### ISSUE THREAD

**Issue by thekaveh** [thekaveh/NNx#231]
**Category:** Configuration and compatibility · **Priority:** P2 · **Size:** M

**Evidence status:** Reproduced input-validation gap with incorrect cadence.

## Outcome and scope

Validate public configuration counts as integers before range arithmetic or layer allocation. Accept `numbers.Integral` values including NumPy integers and normalize accepted values to Python `int`; reject booleans, all floats (including 2.0), strings and nonfinite values. This is count validation, not a scheduler-clock feature or a change to dataset `None` semantics.

## Current behavior / evidence

Current guards check only positive/nonnegative ranges. `accumulate_grad_batches=2.5` therefore reaches modulo arithmetic and yields an unintended five-batch cadence; fractional epochs reach `range` later. Base, Transformer, convolution and MoE params similarly assume integral dimensions. Existing serialization stringifies several lists, so accepted NumPy values need normalization before building immutable lists and state snapshots.

## Repository map

- [`src/nnx/nn/params/nn_params.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/nn/params/nn_params.py#L62) — Base dimensions, hidden dimensions and optional heads.
- [`src/nnx/nn/params/nn_transformer_params.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/nn/params/nn_transformer_params.py#L55) — Transformer dimensions, divisibility and even RoPE head width.
- [`src/nnx/nn/params/nn_conv_params.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/nn/params/nn_conv_params.py#L52) — Convolution counts, nonnegative padding and spatial checks.
- [`src/nnx/nn/params/nn_moe_params.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/nn/params/nn_moe_params.py#L43) — Expert count/top-k bounds.
- [`src/nnx/nn/params/nn_optim_params.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/nn/params/nn_optim_params.py#L51) — Accumulation count before modulo use.
- [`src/nnx/nn/params/nn_scheduler_params.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/nn/params/nn_scheduler_params.py#L31) — Positive length counts versus nonnegative patience/cooldown.
- [`src/nnx/nn/callbacks.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/nn/callbacks.py#L75) — Callback patience is a count, not a real hyperparameter.
- [`tests/test_params_boundary_validation.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/tests/test_params_boundary_validation.py#L53) — Baseline constructor validation and existing error conventions.

## Implementation approach

1. Reuse or add the small private validator proposed by FIX-020, with a count function accepting `numbers.Integral` except `bool`, returning plain `int`, and applying an explicit minimum. Validate before `%`, divisions, `math.isqrt`, list allocation and construction of `_dims`.
2. Inventory positive base input/output/hidden dimensions, optional positive heads; Transformer vocab/layers/model-width/sequence-length/FFN multiplier; convolution channels/kernel/stride/pool and padding ≥0; MoE experts ≥2 and 1≤top_k≤experts. Require `d_model % n_heads == 0` and resulting head dimension even; retain direct RoPE's evenness guard and validate its count inputs too.
3. Apply the same policy to `NNTrainParams.n_epochs`, `NNTrainerParams.n_epochs`, optimizer accumulation (all ≥1), scheduler positive `step_size`/`T_max`/`total_steps`/`warmup_steps` when supplied, and nonnegative scheduler/callback patience/cooldown. Preserve optional None. Do not alter max_lr, factor, threshold or other real-valued knobs. Retain current warmup_steps >0 when explicitly supplied; zero-warmup support is not introduced here.
4. Normalize through `object.__setattr__` inside frozen constructors before serializing. Keep default-omission keys and old valid Python-int states identical. FIX-022 owns tuple length and per-split batch-size resolution; share conventions without broadening this ticket into dataset validation.

## Acceptance criteria

- [ ] Fractional counts, bools, NaN/infinity, and numeric strings fail at public params construction, with named-field errors.
- [ ] Positive integral accumulation values produce the documented windows, including the final short window; epochs and scheduler counts honor their valid zero/positive boundaries.
- [ ] The policy covers base/subclass model dimensions and count fields, training epochs, optimizer accumulation, scheduler counts, and callback patience without treating real-valued hyperparameters as integers.
- [ ] Supported integral scalar types serialize to portable values and round-trip without altering existing valid config hashes.
- [ ] The batch-size-specific dataset ticket remains the owner of split tuple shape and `None` semantics, preventing contradictory validators.

## Validation and completion

Construction oracle with complete setup:

```python
from dataclasses import replace
import numpy as np
import pytest
from nnx import NNOptimParams

base = NNOptimParams.builder().adam(max_lr=1e-3).build()
for bad in [2.5, 2.0, True, "2", float("nan"), float("inf")]:
    with pytest.raises(ValueError, match="accumulate_grad_batches"):
        replace(base, accumulate_grad_batches=bad)
good = replace(base, accumulate_grad_batches=np.int64(2))
assert type(good.accumulate_grad_batches) is int
assert good.state() == replace(base, accumulate_grad_batches=2).state()
```

Create `tests/test_integer_config_validation.py` with constructors spanning the inventory. For valid accumulation 2 over five microbatches, instrument optimizer updates and assert windows end at indices 1, 3 and 4, comparing parameter updates against appropriately normalized reference windows. Validation alone must not rewrite accumulation implementation.

Run existing `uv run --frozen pytest tests/test_params_boundary_validation.py tests/test_nn_transformer_params_builder.py tests/test_conv_nn.py tests/test_nn_moe.py tests/test_trainer_params.py tests/test_callbacks.py tests/test_pass2_f_series.py`. Once created, run `uv run --frozen pytest tests/test_integer_config_validation.py tests/test_params_round_trip.py`. Include zero padding/patience and rejected zero epochs as distinct boundaries.

From the repository root, CI-equivalent setup is `uv sync --frozen --all-extras`. The commands above describe required implementation checks; none were executed during ticket enrichment. Finish with `uv run --frozen ruff check src/ tests/ examples/ scripts/`, `uv run --frozen ruff format --check src/ tests/ examples/ scripts/`, and `uv run --frozen pyright --warnings`. Preserve Python 3.10 compatibility and default-omission serialization; update canonical documentation sources, not generated projections by hand.

## Cascading changes and affected consumers

- **Deserialization and architecture dispatch.** Counts enter through direct builders and through [NNParams.resolve_from_state](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/nn/params/nn_params.py#L164), which selects Transformer, convolution and MoE families. Test malformed integral fields at that dispatch boundary and at NNRun.load, rather than only each dataclass in isolation. Normalize supported NumPy integral values before immutable-list capture and state generation, retaining old Python-int snapshots and default omissions. Do not round fractional YAML values into valid dimensions.
- **Loop and builder consumers.** [Transformer.layers](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/nn/params/nn_transformer_params_builder.py#L39) performs divisibility arithmetic before build; it must validate integer type before that arithmetic so builders and direct constructors reject the same cases. Trainer and NNModel feed n_epochs into range and accumulation into modulo. Scheduler factories consume count fields in native constructors. Preserve distinct zero policies for padding/patience versus epochs/accumulation, and keep None sentinels unchanged. This validation repair must not alter the actual accumulation algorithm or introduce scheduler clocks.
- **Sample execution.** The convolution example reconstructs a saved model through subtype dispatch, and the LM CLI exposes dimensions/heads/layers as integers. Add bounded config-only invalid-input cases plus an executed tiny convolution round trip using NumPy integers, checking layer shape and plain-int serialized values. The sample's main is not currently executed by general smoke discovery, so register a dedicated callable scenario with tiny synthetic input and no full training schedule. Document accepted integer scalar types and constructor-time errors in the relevant builder and shape-helper docstrings.

### Samples and executable coverage

- [`examples/25_conv_classifier.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/examples/25_conv_classifier.py): Expose a tiny config/checkpoint schema helper for NumPy-integer dimensions, retaining normal sample main. **Verify:** Proposed test_conv_sample_integral_roundtrip executes one tiny forward plus checkpoint reconstruction, compares shape and plain-int state, and rejects fractional counts before network allocation.

Update `examples/README.md` with each affected sample’s purpose, extras, input/resource needs and bounded invocation. Preserve import-without-main behavior; add explicit import and execution registration in `tests/test_examples_smoke.py` for new unnumbered scripts. Existing discovery covers only `[0-9][0-9]_*.py`, and only selected scripts execute by default.

### Repository, site and wiki propagation

Edit these canonical sources; the two other columns are generated projections, not independently maintained pages. Proposed guide slugs below are explicit suggestions: reuse the prerequisite’s registered slug when it already exists.

| Canonical repository source | Generated site Markdown | Generated wiki Markdown | Required content change |
|---|---|---|---|
| [`docs/concepts.md`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/docs/concepts.md) | `generated/site/Concepts.md` | `generated/wiki/Concepts.md` | Document nonboolean Integral count policy, NumPy normalization, and distinct allowed zero boundaries. |
| [`docs/lm.md`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/docs/lm.md) | `generated/site/Language-modeling.md` | `generated/wiki/Language-modeling.md` | State positive integer dimensions/layers, head divisibility and even RoPE head width at builder and direct construction. |

Update affected public docstrings/API catalog and regenerate `docs/api.md`; add the scoped changelog entry and reconcile changed top-level claims. Register new guides in `docs/manifest.yaml` with unique slugs/numbers and consistent headings/navigation. Repository-link validation uses Git-tracked paths, so newly linked source files must be tracked/staged before that check. Generated pages are never edited directly.

After source/API edits, run from the repository root in the existing frozen all-extras environment (regenerate/review the lock first only if this ticket changes dependencies):

```bash
uv run --frozen python -m scripts.docs.build_api_reference
uv run --frozen python -m scripts.docs.build_docs
uv run --frozen mkdocs build --strict
uv run --frozen python -m scripts.docs.build_wiki --check
uv run --frozen pytest tests/test_docs_projection.py tests/test_wiki_projection.py
```

If README changed, run `uv run --frozen python -m scripts.docs.build_pypi_readme` and its `--check` form. If canonical diagram HTML changed, run `uv run --frozen python -m scripts.docs.extract_architecture_svg` before the docs build and verify with `--check`. Compare the changed repository passages with rendered `site/` and wiki projections: values, signatures, limitations, examples, links/anchors and assets must agree. Deterministic generation alone is not a freshness or live-publication check.

**Delivery evidence:** record revision, executed checks and changed page mapping. `.github/workflows/docs.yml` builds PRs/main/develop and publishes Pages/wiki on main pushes. Verify served content after that authorized publication; record deployment failure/skew as pending follow-through. Local rendering is not live publication, and PR preparation requires no manual wiki/site push.

### Additional acceptance criteria — cascading completion

- [ ] Subtype reconstruction through NNParams.resolve_from_state and NNRun.load rejects fractional or boolean count fields with the originating field name and retains valid historical integer states.
- [ ] Transformer builder layers validation rejects nonintegral heads/d_model before divisibility arithmetic and matches the direct constructor policy.
- [ ] A registered bounded execution of the convolution sample round-trips accepted NumPy-integer dimensions into plain-int metadata and an identical reconstructed output shape.
- [ ] The issue-specific changes in `docs/concepts.md`, `docs/lm.md`, affected API reference, sample catalog and changelog are reflected consistently in repository, generated/rendered site and wiki; documented commands/examples and links pass the checks above, and delivery evidence distinguishes the reviewed revision from any pending live publication.

**Coordination:** The traced downstream changes belong in this ticket. Shared-file coordination and optional combinations do not add a hard prerequisite.

Additional nonblocking consumers/coordination: [FIX-022](https://github.com/thekaveh/NNx/issues/232), [FEAT-014](https://github.com/thekaveh/NNx/issues/236).

## Delivery order and dependencies

- **Execution order:** 32 / 67 (priority-aware topological order).
- **Dependency wave:** 1. Items in the same wave may proceed independently when their prerequisites are satisfied.
- **Blocked by:** None within this backlog.
- **Related / optional integration:** [FEAT-004](https://github.com/thekaveh/NNx/issues/214), [FIX-020](https://github.com/thekaveh/NNx/issues/230), [FIX-022](https://github.com/thekaveh/NNx/issues/232), [FEAT-014](https://github.com/thekaveh/NNx/issues/236). These links do not block this ticket.

**Dependency rationale.** Independent repair of current behavior. Coordinate with FIX-020 and FIX-022; scheduler units remain FEAT-014.

Reviewed against [`0c707eb`](https://github.com/thekaveh/NNx/commit/0c707eb5b80d2173843e60102ba326e4f997749d) on 2026-09-20. Acceptance criteria are prospective; this issue does not claim an implementation or new test pass.

**Comment by thekaveh** [comment]
Resolved in #299 (develop, merged `ecc336d`) and #300 (main, merged `d4fca9e`).

**Root cause.** Every count guard checked only a range (`< 1`, `<= 0`, `< 0`), so `accumulate_grad_batches=2.5` reached `batch_idx % 2.5` and silently stepped every five batches, fractional epochs reached `range()` at train time, `True` passed every guard as `1`, NaN/inf sailed through the comparisons, and NumPy integers were kept as-is — a `numpy.int64` count made `yaml.safe_dump` of `run.yaml` raise `RepresenterError`.

**What changed.**
- New `require_count(value, field, *, owner, minimum, maximum, exclusive_min, exclusive_max, domain_message) -> int` in `src/nnx/_validation.py` (the FIX-020 sibling): accepts any `numbers.Integral` except `bool` — NumPy integers included — and returns a plain `int`; rejects every float (`2.0` included), strings, `None`, NaN/±inf and out-of-range values with a `ValueError` naming owner, field and domain. Established out-of-range wordings are kept via `domain_message`.
- Wired, normalizing through `object.__setattr__` before immutable lists, `_dims`, `%` / `isqrt` / `range` arithmetic and `state()`: `NNParams.input_dim` / `output_dim` / each `hidden_dims` entry / `n_heads`; `NNTransformerParams.vocab_size` / `n_layers` / `d_model` / `max_seq_len` / `ffn_mult` plus a new even-head-width check (`d_model // n_heads`, required by RoPE) after divisibility, mirrored in `NNTransformerParams.builder().layers()` *before* its modulo; `NNConvParams.conv_channels` entries / `in_channels` / `kernel_size` / `stride` / `pool_size` > 0 and `padding` ≥ 0 before the square-image arithmetic; `NNMoEParams.num_experts` ≥ 2 / `top_k` > 0; `NNOptimParams.accumulate_grad_batches` ≥ 1; `NNSchedulerParams.patience` / `cooldown` ≥ 0 and `step_size` / `T_max` / `total_steps` / `warmup_steps` > 0 when set (`None` untouched); `NNTrainParams.n_epochs` / `NNTrainerParams.n_epochs` ≥ 1; `EarlyStopping(patience)` ≥ 0; direct `RoPE(dim, max_seq_len)` > 0 with the evenness guard kept. Real-valued knobs, `None` sentinels, zero policies and the `state()` of valid Python-int configs are unchanged; `np.int64` configs now serialize and hash identically to their plain-int twins.
- Example 25: bounded `conv_integral_schema_roundtrip()` (NumPy-integer dims → plain-int state → YAML → `resolve_from_state` → identical rebuilt forward; fractional/boolean counts rejected before allocation), registered in the smoke harness and called from `main()`. Docs: `docs/concepts.md` §2.3, `docs/lm.md`, `NNConvParams` / builder `.layers()` docstrings, `examples/README.md`, `CHANGELOG.md [Unreleased]`, regenerated `docs/api.md`.

**Acceptance criteria → tests** (`tests/test_integer_config_validation.py`, 186 tests):
- Named-field rejection of fractions, `2.0`, bools, strings, NaN, inf across 30 constructor sites (180 cases) — `test_non_integral_counts_are_rejected_with_field_name`.
- Accumulation 2 over 5 microbatches steps after batches 1, 3, 4 with parameters equal to an explicit normalized-window reference (the cadence, not the algorithm, changed) — `test_valid_accumulation_windows_end_at_documented_indices`.
- Zero padding / patience / cooldown valid, zero epochs / accumulation / dims / step counts rejected, `None` sentinels kept — `test_zero_and_positive_boundaries_keep_their_distinct_policies`.
- NumPy integers → `type(...) is int`, identical `state()`, YAML-safe (ticket oracle included) — `test_numpy_integers_normalize_to_plain_int_and_serialize`.
- `resolve_from_state`, `NNTrainParams.from_state`, `NNRun.load` on a malformed `run.yaml` reject with the field name — `test_builder_and_from_state_routes_reject_the_same_cases`, `test_malformed_count_in_run_yaml_fails_at_load`.
- Builder `.layers()` rejects non-integral `d_model` before divisibility; even head width enforced at both routes — `test_transformer_head_dim_must_be_even_and_integral`.
- Dataset `batch_sizes` semantics untouched (FIX-022 owns them).

**Checks (both PRs, local + CI green: lint-and-test 3.10–3.14, pip-audit, CodeQL, docs build):**
```
uv run --frozen pytest -q                                    → 2007 passed, 1 skipped, 2 deselected (libcairo docs-projection tests, CI runs them)
uv run --frozen ruff check src/ tests/ examples/ scripts/    → All checks passed
uv run --frozen ruff format --check …                        → 256 files already formatted
uv run --frozen pyright --warnings                           → 0 errors, 0 warnings
build_api_reference / build_docs / mkdocs build --strict / build_wiki --check / projection tests → pass
```
Live site/wiki publication runs from the main push (`docs.yml`); not re-verified here.
